core/services/offchainreporting/OCRContract.go

```diff
@@ -0,0 +1,228 @@
+	if err != nil {
+		return o, errors.Wrap(err, "could not instantiate NewOffchainAggregatorFilterer")
+	}
+
+	contractCaller, err := offchainaggregator.NewOffchainAggregatorCaller(address, ethClient)
+	if err != nil {
+		return o, errors.Wrap(err, "could not instantiate NewOffchainAggregatorCaller")
+	}
+
+	contractABI, err := abi.JSON(strings.NewReader(offchainaggregator.OffchainAggregatorABI))
+	if err != nil {
+		return o, errors.Wrap(err, "could not get contract ABI JSON")
+	}
+
+	return &OCRContract{
+		contractFilterer: contractFilterer,
+		contractCaller:   contractCaller,
+		contractAddress:  address,
+		logBroadcaster:   logBroadcaster,
+		ethClient:        ethClient,
+		configSetChan:    make(chan offchainaggregator.OffchainAggregatorConfigSet),
+		jobID:            jobID,
+		transmitter:      transmitter,
+		contractABI:      contractABI,
+	}, nil
+}
+
+func (ra *OCRContract) Transmit(ctx context.Context, report []byte, rs, ss [][32]byte, vs [32]byte) error {
+	payload, err := ra.contractABI.Pack("transmit", report, rs, ss, vs)
+	if err != nil {
+		return errors.Wrap(err, "abi.Pack failed")
+	}
+
+	return errors.Wrap(ra.transmitter.CreateEthTransaction(ctx, ra.contractAddress, payload), "failed to send Eth transaction")
+}
+
+func (ra *OCRContract) SubscribeToNewConfigs(context.Context) (ocrtypes.ContractConfigSubscription, error) {
+	sub := &OCRContractConfigSubscription{
+		make(chan ocrtypes.ContractConfig),
+		ra,
+		sync.Mutex{},
+		false,
+	}
+	connected := ra.logBroadcaster.Register(ra.contractAddress, sub)
+	if !connected {
+		return nil, errors.New("Failed to register with logBroadcaster")
+	}
+
+	return sub, nil
+}
+
+func (ra *OCRContract) LatestConfigDetails(ctx context.Context) (changedInBlock uint64, configDigest ocrtypes.ConfigDigest, err error) {
+	opts := bind.CallOpts{Context: ctx, Pending: false}
+	result, err := ra.contractCaller.LatestConfigDetails(&opts)
+	if err != nil {
+		return 0, configDigest, errors.Wrap(err, "error getting LatestConfigDetails")
+	}
+	return uint64(result.BlockNumber), ocrtypes.BytesToConfigDigest(result.ConfigDigest[:]), err
+}
+
+// Conform OCRContract to LogListener interface
+type OCRContractConfigSubscription struct {
+	ch       chan ocrtypes.ContractConfig
+	ra       *OCRContract
+	mutex    sync.Mutex
+	chClosed bool
+}
+
+func (sub *OCRContractConfigSubscription) OnConnect() {}
+func (sub *OCRContractConfigSubscription) OnDisconnect() {
+	sub.mutex.Lock()
+	defer sub.mutex.Unlock()
+
+	if !sub.chClosed {
+		sub.chClosed = true
+		close(sub.ch)
+	}
+}
+func (sub *OCRContractConfigSubscription) HandleLog(lb eth.LogBroadcast, err error) {
+	topics := lb.Log().RawLog().Topics
+	if len(topics) == 0 {
+		return
+	}
+	switch topics[0] {
+	case OCRContractConfigSet:
+		configSet, err := sub.ra.contractFilterer.ParseConfigSet(lb.Log().RawLog())
+		if err != nil {
+			panic(err)
+		}
+		configSet.Raw = lb.Log().RawLog()
```

Comment:
Can we insert a check here that the address of the OCRContract matches the address in the raw log? This is to prevent a malicious geth node from lying about the address, which could cause safety issues.